### ISSUE TITLE
SQL Processors are MySQL-specific

### ISSUE THREAD

**Issue by greggroth** [dailyburn/ratchet#21]
The `OnDupKeyUpdate` option uses the MySQL-specific `ON DUPLICATE KEY UPDATE` update syntax, but that detail isn't mentioned anywhere in the documentation.  I hope to work on a PR to extend it to also support PostgreSQL 9.5's `ON CONFLICT` syntax, but in the mean time I want to open a PR to document the features that are MySQL-specific.  Other than `OnUpKeyUpdate`, are there any other places that could benefit from calling this out in the documentation?

**Comment by sturgill** [comment]
That's actually the only place right now where we're even trying to be cute about generating SQL. Everywhere else expects pre-crafted queries.

**Comment by greggroth** [comment]
:+1:   I'll see if it's reasonable to update that to support postgres or if it would better to create a new processor that accepts a custom insert query as an argument so the pg-specific query can be injected that way.

**Comment by sturgill** [comment]
@greggroth: Take a look at [this feature branch](https://github.com/dailyburn/ratchet/tree/feature%2321-postgres-upsert) I just pushed. I think an approach like this should be pretty reasonable.

I'm not familiar with PostgreSQL, so I didn't actually wire up the SQL generation. But if you want to take a stab at filling out that func and submit a PR against that branch, it would be lovely.

**Comment by sturgill** [comment]
@greggroth: We are actively looking here to start using Redshift as an internal data lake. The upside is that it appears that you can use standard PostgreSQL drivers to connect to Redshift (meaning we have a vested interest in supporting PostgreSQL as a first-class citizen).

In addition to `ON CONFLICT` PostgreSQL wants prepared statements to have `$1` type syntax for value placeholders (instead of MySQL's `?`). As a result of this, it made the most sense to split `SQLWriter` into `MySQLWriter` and `PostgreSQLWriter`.

This will be a breaking change and will be released with a v2.0.x tag. I'd love your feedback on [the PostgreSQL feature branch](https://github.com/dailyburn/ratchet/tree/feature%2321-postgres-upsert).

There are a couple of other requests we might want to slip into a new release as well (#23 and #30  come to mind).

**Comment by llonchj** [comment]
+1 @sturgill looks right.

The implementation for `database/sql` placeholders depends on the driver: mysql `?`, postgresql `$1, $2`, sqlite supports `$1` or `?`, oracle `:col_name`.

**Comment by sturgill** [comment]
Correct. The issue is that the existing `SQLWriter` hard-coded the MySQL expectation (`?`). A complete separation of MySQLWriter and PostgreSQLWriter seems like the best approach. A little copying something something.

**Comment by llonchj** [comment]
It seems the SQLWriter approach should be driver specific, unless support for dialects and placeholders are in place.

Build a plain SQL query without placeholders will solve only part of the situation as UPSERT has database specific syntax.

**Comment by sturgill** [comment]
Splitting `SQLWriter` into two components should address both UPSERT and placeholder nuances.
